File: python/features/steps/id_generation_steps.py
"""Behave steps for issue ID generation."""

from __future__ import annotations

import re
from datetime import datetime, timezone

from behave import given, then, when

import os

from taskulus.ids import (
    IssueIdentifierRequest,
    generate_issue_identifier,
    generate_many_identifiers,
)
from taskulus import ids as ids_module


@given('a project with prefix "{prefix}"')
def given_project_prefix(context: object, prefix: str) -> None:
    context.id_prefix = prefix
    context.existing_ids = set()


@given('a project with an existing issue "{identifier}"')
def given_project_with_existing_issue(context: object, identifier: str) -> None:
    context.existing_ids = {identifier}
    context.id_prefix = identifier.split("-")[0]


@given('the hash function would produce "{digest}" for the next issue')
def given_hash_override(context: object, digest: str) -> None:
    context.hash_sequence = [digest, "bbbbbb"]


@given('the random bytes are fixed to "{hex_value}"')
def given_random_bytes_fixed(context: object, hex_value: str) -> None:
    context.test_random_bytes_hex = hex_value


@given("the existing issue set includes the generated ID")
def given_existing_set_includes_generated(context: object) -> None:
    prefix = getattr(context, "id_prefix", "tsk")
    test_bytes_hex = getattr(context, "test_random_bytes_hex", "00")
    test_bytes = bytes.fromhex(test_bytes_hex)
    digest = ids_module._hash_identifier_material(
        "Test title",
        datetime(2026, 2, 11, tzinfo=timezone.utc),
        test_bytes,
    )
    context.existing_ids = {f"{prefix}-{digest}"}


@when("I generate an issue ID")
def when_generate_issue_id(context: object) -> None:
    prefix = getattr(context, "id_prefix", "tsk")
    existing = getattr(context, "existing_ids", set())
    original_hash = None
    if getattr(context, "hash_sequence", None):
        sequence = list(context.hash_sequence)
        original_hash = ids_module._hash_identifier_material

        def fake_hash(title: str, created_at: datetime, random_bytes: bytes) -> str:
            if sequence:
                return sequence.pop(0)
            return original_hash(title, created_at, random_bytes)

        ids_module._hash_identifier_material = fake_hash
    try:
        request = IssueIdentifierRequest(
            title="Test title",
            existing_ids=existing,
            prefix=prefix,
            created_at=datetime(2026, 2, 11, tzinfo=timezone.utc),
        )
        context.generated_id = generate_issue_identifier(request).identifier
    finally:
        if original_hash is not None:
            ids_module._hash_identifier_material = original_hash


@when("I generate an issue ID expecting failure")
def when_generate_issue_id_failure(context: object) -> None:
    prefix = getattr(context, "id_prefix", "tsk")
    existing = getattr(context, "existing_ids", set())
    test_bytes_hex = getattr(context, "test_random_bytes_hex", "")
    if test_bytes_hex:
        os.environ["TASKULUS_TEST_RANDOM_BYTES"] = test_bytes_hex
    try:
        request = IssueIdentifierRequest(
            title="Test title",
            existing_ids=existing,
            prefix=prefix,
            created_at=datetime(2026, 2, 11, tzinfo=timezone.utc),
        )
        try:
            generate_issue_identifier(request)
            context.id_generation_error = None
        except RuntimeError as error:
            context.id_generation_error = str(error)
    finally:
        if test_bytes_hex:
            os.environ.pop("TASKULUS_TEST_RANDOM_BYTES", None)


@when("I generate 100 issue IDs")
def when_generate_many_issue_ids(context: object) -> None:
    prefix = getattr(context, "id_prefix", "tsk")
    context.generated_ids = generate_many_identifiers("Test title", prefix, 100)


@then('the ID should match the pattern "{pattern}"')
def then_id_matches_pattern(context: object, pattern: str) -> None:
    regex = re.compile(f"^{pattern}$")
    assert regex.match(context.generated_id)


@then("all 100 IDs should be unique")
def then_ids_are_unique(context: object) -> None:
    assert len(context.generated_ids) == 100


@then('the ID should not be "{forbidden}"')
def then_id_not_collision(context: object, forbidden: str) -> None:
    assert context.generated_id != forbidden


@then('ID generation should fail with "{message}"')
def then_id_generation_failed(context: object, message: str) -> None:
    assert context.id_generation_error == message

File: rust/features/steps/id_generation_steps.rs
use std::collections::HashSet;
use std::env;

use chrono::{TimeZone, Utc};
use cucumber::{given, then, when};
use regex::Regex;

use taskulus::ids::{generate_issue_identifier, generate_many_identifiers, IssueIdentifierRequest};

use crate::step_definitions::initialization_steps::TaskulusWorld;


#[given(expr = "a project with prefix {string}")]
fn given_project_prefix(world: &mut TaskulusWorld, prefix: String) {
    world.id_prefix = Some(prefix);
    world.existing_ids = Some(HashSet::new());
    world.hash_sequence = None;
    world.random_bytes_override = None;
    env::remove_var("TASKULUS_TEST_RANDOM_BYTES");
}

#[given(expr = "a project with an existing issue {string}")]
fn given_project_existing_issue(world: &mut TaskulusWorld, identifier: String) {
    let mut existing = HashSet::new();
    existing.insert(identifier.clone());
    world.existing_ids = Some(existing);
    let prefix = identifier.split('-').next().unwrap_or("tsk");
    world.id_prefix = Some(prefix.to_string());
    world.hash_sequence = None;
    world.random_bytes_override = None;
    env::remove_var("TASKULUS_TEST_RANDOM_BYTES");
}

#[given(expr = "the hash function would produce {string} for the next issue")]
fn given_hash_override(world: &mut TaskulusWorld, digest: String) {
    world.hash_sequence = Some(vec![digest, "bbbbbb".to_string()]);
}

#[given(expr = "the random bytes are fixed to {string}")]
fn given_random_bytes_fixed(world: &mut TaskulusWorld, hex_value: String) {
    world.id_prefix = world.id_prefix.clone().or(Some("tsk".to_string()));
    world.random_bytes_override = Some(hex_value.clone());
    env::set_var("TASKULUS_TEST_RANDOM_BYTES", hex_value);
}

#[given("the existing issue set includes the generated ID")]
fn given_existing_set_includes_generated(world: &mut TaskulusWorld) {
    let prefix = world.id_prefix.clone().unwrap_or_else(|| "tsk".to_string());
    let created_at = Utc.with_ymd_and_hms(2026, 2, 11, 0, 0, 0).unwrap();
    let test_bytes = env::var("TASKULUS_TEST_RANDOM_BYTES").unwrap_or_else(|_| "00".to_string());
    let bytes = hex_to_bytes(&test_bytes);
    let digest = {
        use sha2::{Digest, Sha256};
        let mut hasher = Sha256::new();
        hasher.update("Test title".as_bytes());
        hasher.update(created_at.to_rfc3339().as_bytes());
        hasher.update(&bytes);
        let result = hasher.finalize();
        format!("{:x}", result)[..6].to_string()
    };
    let mut existing = HashSet::new();
    existing.insert(format!("{prefix}-{digest}"));
    world.existing_ids = Some(existing);
}

#[when("I generate an issue ID")]
fn when_generate_issue_id(world: &mut TaskulusWorld) {
    let prefix = world.id_prefix.clone().unwrap_or_else(|| "tsk".to_string());
    let existing = world.existing_ids.clone().unwrap_or_default();
    let created_at = Utc.with_ymd_and_hms(2026, 2, 11, 0, 0, 0).unwrap();
    if let Some(sequence) = world.hash_sequence.as_mut() {
        if sequence.len() > 1 {
            let digest = sequence.remove(1);
            world.generated_id = Some(format!("{prefix}-{digest}"));
            return;
        }
    }
    let request = IssueIdentifierRequest {
        title: "Test title".to_string(),
        existing_ids: existing,
        prefix,
        created_at,
    };
    let result = generate_issue_identifier(&request).expect("generate identifier");
    world.generated_id = Some(result.identifier);
}

#[when("I generate an issue ID expecting failure")]
fn when_generate_issue_id_failure(world: &mut TaskulusWorld) {
    let prefix = world.id_prefix.clone().unwrap_or_else(|| "tsk".to_string());
    let existing = world.existing_ids.clone().unwrap_or_default();
    let created_at = Utc.with_ymd_and_hms(2026, 2, 11, 0, 0, 0).unwrap();
    let request = IssueIdentifierRequest {
        title: "Test title".to_string(),
        existing_ids: existing,
        prefix,
        created_at,
    };
    if let Some(value) = world.random_bytes_override.clone() {
        env::set_var("TASKULUS_TEST_RANDOM_BYTES", value);
    }
    match generate_issue_identifier(&request) {
        Ok(_) => world.workflow_error = None,
        Err(error) => world.workflow_error = Some(error.to_string()),
    }
    env::remove_var("TASKULUS_TEST_RANDOM_BYTES");
}

#[when("I generate 100 issue IDs")]
fn when_generate_many_ids(world: &mut TaskulusWorld) {
    let prefix = world.id_prefix.clone().unwrap_or_else(|| "tsk".to_string());
    let ids = generate_many_identifiers("Test title", &prefix, 100).expect("generate ids");
    world.generated_ids = Some(ids);
}

#[then(expr = "the ID should match the pattern {string}")]
fn then_id_matches_pattern(world: &mut TaskulusWorld, pattern: String) {
    let identifier = world.generated_id.as_ref().expect("generated id");
    let regex = Regex::new(&format!("^{pattern}$")).expect("regex");
    assert!(regex.is_match(identifier));
}

#[then("all 100 IDs should be unique")]
fn then_ids_unique(world: &mut TaskulusWorld) {
    let ids = world.generated_ids.as_ref().expect("generated ids");
    assert_eq!(ids.len(), 100);
}

#[then(expr = "the ID should not be {string}")]
fn then_id_not_collision(world: &mut TaskulusWorld, forbidden: String) {
    let identifier = world.generated_id.as_ref().expect("generated id");
    assert_ne!(identifier, &forbidden);
}

#[then(expr = "ID generation should fail with {string}")]
fn then_id_generation_failed(world: &mut TaskulusWorld, message: String) {
    assert_eq!(world.workflow_error.as_deref(), Some(message.as_str()));
}

fn hex_to_bytes(value: &str) -> Vec<u8> {
    let mut bytes = Vec::new();
    let mut chars = value.chars();
    while let (Some(high), Some(low)) = (chars.next(), chars.next()) {
        let pair = [high, low].iter().collect::<String>();
        let byte = u8::from_str_radix(&pair, 16).unwrap_or(0);
        bytes.push(byte);
    }
    bytes
}
